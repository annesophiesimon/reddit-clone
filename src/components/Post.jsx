import Vote from './Vote';

const Post = (post) => {
  const { ups, title, thumbnail, author, dateCreated, nbComs } = post;
  return (
    <>
      <div className="bg-white flex py-5 shadow-lg rounded-xl m-4 md:w-5/6">
        <div className="flex w-full flex-col">
          <div className="flex items-start mx-2">
            <Vote ups={ups} />
            <section className="text-center">
              <article className="w-full">
                <h1 className="mb-2"> {title}</h1>
                <div className="flex items-center justify-center mb-8">
                  <img className="w-6/12" alt={title} src={thumbnail}></img>
                </div>
                <hr />
                <p className="mt-8">
                  {author} - {dateCreated} - {nbComs}
                </p>
              </article>
            </section>
          </div>
        </div>
      </div>
    </>
  );
};

export default Post;
