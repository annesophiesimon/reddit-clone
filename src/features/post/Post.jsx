/* eslint-disable react/prop-types */
import Vote from '../../components/Vote';
import { convertTime, checkURL } from '../../helpers';
import { fetchPosts } from './postSlice';
import { useAppSelector, useAppDispatch } from '../../app/hooks';
import { useEffect } from 'react';

const Post = (props) => {
  const { subreddit } = props;
  const { posts, isLoading } = useAppSelector((state) => state.post);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(fetchPosts(subreddit));
  }, [subreddit]);
  if (isLoading) {
    return <div>Data loading</div>; // Add spinner
  }
  return (
    <>
      {posts.length > 0 ? (
        posts
          .filter((post) => checkURL(post.url))
          .map((post, index) => (
            <div key={index} className="container mx-auto  bg-white">
              <div className="flex py-5 shadow-lg border border-grey-50 rounded-xl max-w-[700px] mb-10">
                <div className="flex flex-col">
                  <div className="flex items-start">
                    <section>
                      <article>
                        <div className="px-5 mb-5">
                          <h1 className="font-bold text-[#212427] text-sm">{post.author}</h1>
                          <p className="text-slate-500 text-xs">{convertTime(post.created_utc)}</p>
                          <h1 className="my-4  text-[#212427]"> {post.title}</h1>
                        </div>
                        <div className="flex items-center justify-center mb-8">
                          <img
                            className="justify-items-center"
                            alt={post.title}
                            src={post.url}></img>
                        </div>
                        <hr className="mx-20 mb-10" />
                      </article>
                    </section>
                  </div>
                  <div className="flex justify-between text-sm text-slate-500 px-5">
                    <Vote ups={post.ups} />
                    <button>
                      <p>
                        {post.num_comments} {post.num_comments > 1 ? `comments` : `comment`}
                      </p>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          ))
      ) : (
        <p> There is no post yet </p>
      )}
    </>
  );
};

export default Post;
