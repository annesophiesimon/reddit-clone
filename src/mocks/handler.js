import { rest } from 'msw';
import { setupServer } from 'msw/lib/node';

import {
  resultMockSubreddit,
  resultMockSewing,
  resultMockCrossfit,
  resultMockBachata,
  resultMockEmptyPost
} from './result';

const API_ROOT = 'https://www.reddit.com';
export const handlers = [
  rest.get(`${API_ROOT}/subreddits.json`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(resultMockSubreddit), ctx.delay(50));
  }),
  rest.get(`${API_ROOT}/r/sewing.json`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(resultMockSewing), ctx.delay(150));
  }),
  rest.get(`${API_ROOT}/r/crossfit.json`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(resultMockCrossfit), ctx.delay(150));
  }),
  rest.get(`${API_ROOT}/search.json`, (req, res, ctx) => {
    const q = req.url.searchParams.get('q');
    if (q === 'bachata') {
      return res(ctx.status(200), ctx.json(resultMockBachata), ctx.delay(150));
    } else {
      return res(ctx.status(200), ctx.json(resultMockEmptyPost), ctx.delay(150));
    }
  })
];

export const server = setupServer(...handlers);

server.listen();
