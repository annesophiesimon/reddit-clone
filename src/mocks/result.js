export const resultMockSubreddit = {
  data: {
    children: [
      { data: { id: 1, display_name_prefixed: 'r/home', display_name: 'Home' } },
      { data: { id: 2, display_name_prefixed: 'r/sewing', display_name: 'Sewing' } },
      { data: { id: 3, display_name_prefixed: 'r/basket', display_name: 'Basket' } },
      { data: { id: 4, display_name_prefixed: 'r/crossfit', display_name: 'Crossfit' } },
      { data: { id: 5, display_name_prefixed: 'r/tech', display_name: 'Tech' } }
    ]
  }
};

export const resultMockSewing = {
  data: {
    children: [
      {
        data: {
          id: 1,
          ups: 26,
          title: 'My completed coat',
          url: 'image.jpg',
          author: 'Kevin',
          created_utc: 36367,
          num_comments: 256,
          permalink: 'r/sewing/comments/1/my_completed_coat'
        }
      },
      {
        data: {
          id: 2,
          ups: 33,
          title: 'how to sew a zipper',
          url: 'image.jpg',
          author: 'Bernard',
          created_utc: 23578899,
          num_comments: 346,
          permalink: 'r/sewing/comments/2/how_to_sew_a_zipper'
        }
      }
    ]
  }
};

export const resultMockCrossfit = {
  data: {
    children: [
      {
        data: {
          id: 1,
          ups: 26,
          title: 'How to improve my muscle up',
          url: 'image.jpg',
          author: 'Katarina',
          created_utc: 36367,
          num_comments: 78,
          permalink: 'r/crossfit/comments/1/how_to_improve_my_muscle_up'
        }
      }
    ]
  }
};

export const resultMockBachata = {
  data: {
    children: [
      {
        data: {
          id: 1,
          ups: 33,
          title: 'Bachata',
          url: 'image.jpg',
          author: 'Lola',
          created_utc: 3636,
          num_comments: 22,
          permalink: 'r/bachata/comments/1/class_in_vegas'
        }
      }
    ]
  }
};

export const resultMockEmptyPost = {
  data: {
    children: []
  }
};

export const resultMockComments = {
  data: {
    children: [
      {
        data: {
          id: 1,
          author: 'Joel',
          created_utc: 36368,
          body: 'I like this pattern'
        }
      }
    ]
  }
};

export const resultsMock = [
  resultMockSubreddit,
  resultMockSewing,
  resultMockBachata,
  resultMockEmptyPost,
  resultMockComments
];
