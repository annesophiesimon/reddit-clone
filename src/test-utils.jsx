// create utils folder
import React from 'react';
import store from './App/store';
import { Provider } from 'react-redux';

export const renderWithProviders = (component) => {
  <Provider store={store}>{component}</Provider>;
};
