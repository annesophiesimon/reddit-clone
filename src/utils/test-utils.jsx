import React from 'react';
import PropTypes from 'prop-types';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    Wrapper.propTypes = {
      children: PropTypes.string.isRequired
    };
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
