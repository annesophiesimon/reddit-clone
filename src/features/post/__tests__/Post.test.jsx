import React from 'react';
import { screen } from '@testing-library/react';
import Post from '../Post';
import { renderWithProviders } from '../../../utils/test-utils';
import { server } from '../../../mocks/handler';

// Enable API mocking before tests.
beforeEach(() => server.listen());

// Disable API mocking after the tests are done.
afterAll(() => server.close());

describe('Post', () => {
  it('renders correctly', () => {
    const post = renderWithProviders(<Post subreddit="r/sewing" />);
    expect(post.container).toMatchSnapshot();
  });

  it('should show loading state', async () => {
    renderWithProviders(<Post subreddit="r/sewing" />);

    expect(screen.getByText(/There is no post yet/i)).toBeInTheDocument();
  });
  it('should render post result', async () => {
    renderWithProviders(<Post subreddit="r/sewing" />);
    expect(await screen.findByText(/kevin/i)).toBeInTheDocument();
  });
});

// add test for comment?
// add test on click on categories
// await userEvent.click(screen.getByRole('button', { name: /Crossfit/i }));

// expect(await screen.findByText(/How to improve my muscle up/i)).toBeInTheDocument();
