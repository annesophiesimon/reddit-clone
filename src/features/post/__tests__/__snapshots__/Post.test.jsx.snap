// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post renders correctly 1`] = `
<div>
  <p>
     There is no post yet 
  </p>
</div>
`;
